File: sonet-client/android/app/src/main/java/xyz/sonet/app/viewmodels/VideoFeedViewModel.kt
package xyz.sonet.app.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import xyz.sonet.app.grpc.SonetGRPCClient
import xyz.sonet.app.grpc.SonetConfiguration
import xyz.sonet.app.grpc.proto.*
import java.util.*

class VideoFeedViewModel(application: Application) : AndroidViewModel(application) {
    
    // MARK: - State Properties
    private val _videos = MutableStateFlow<List<VideoItem>>(emptyList())
    val videos: StateFlow<List<VideoItem>> = _videos.asStateFlow()
    
    private val _currentVideoIndex = MutableStateFlow(0)
    val currentVideoIndex: StateFlow<Int> = _currentVideoIndex.asStateFlow()
    
    private val _selectedTab = MutableStateFlow(VideoTab.FOR_YOU)
    val selectedTab: StateFlow<VideoTab> = _selectedTab.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()
    
    private val _isLiked = MutableStateFlow(false)
    val isLiked: StateFlow<Boolean> = _isLiked.asStateFlow()
    
    private val _isFollowing = MutableStateFlow(false)
    val isFollowing: StateFlow<Boolean> = _isFollowing.asStateFlow()
    
    private val _showComments = MutableStateFlow(false)
    val showComments: StateFlow<Boolean> = _showComments.asStateFlow()
    
    private val _showShare = MutableStateFlow(false)
    val showShare: StateFlow<Boolean> = _showShare.asStateFlow()
    
    private val _currentVideo = MutableStateFlow<VideoItem?>(null)
    val currentVideo: StateFlow<VideoItem?> = _currentVideo.asStateFlow()
    
    // MARK: - Private Properties
    private val grpcClient = SonetGRPCClient(application, SonetConfiguration.development)
    private var currentPage = 0
    private val pageSize = 10
    
    // MARK: - Initialization
    init {
        loadVideos()
    }
    
    // MARK: - Public Methods
    fun loadVideos() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            
            try {
                val response = grpcClient.getVideos(
                    tab = selectedTab.value.grpcType,
                    page = currentPage,
                    pageSize = pageSize
                )
                
                val newVideos = response.videosList.map { VideoItem.from(it) }
                
                if (currentPage == 0) {
                    _videos.value = newVideos
                } else {
                    _videos.value = _videos.value + newVideos
                }
                
                if (_videos.value.isNotEmpty()) {
                    _currentVideo.value = _videos.value[0]
                }
                
                currentPage++
            } catch (e: Exception) {
                _error.value = "Failed to load videos: ${e.localizedMessage}"
            }
            
            _isLoading.value = false
        }
    }
    
    fun switchTab(tab: VideoTab) {
        _selectedTab.value = tab
        currentPage = 0
        _videos.value = emptyList()
        _currentVideoIndex.value = 0
        loadVideos()
    }
    
    fun nextVideo() {
        if (_currentVideoIndex.value < _videos.value.size - 1) {
            _currentVideoIndex.value++
            _currentVideo.value = _videos.value[_currentVideoIndex.value]
            updateVideoState()
        } else {
            // Load more videos if needed
            loadVideos()
        }
    }
    
    fun previousVideo() {
        if (_currentVideoIndex.value > 0) {
            _currentVideoIndex.value--
            _currentVideo.value = _videos.value[_currentVideoIndex.value]
            updateVideoState()
        }
    }
    
    fun selectVideo(index: Int) {
        if (index >= 0 && index < _videos.value.size) {
            _currentVideoIndex.value = index
            _currentVideo.value = _videos.value[index]
            updateVideoState()
        }
    }
    
    fun toggleLike() {
        val video = _currentVideo.value ?: return
        
        _isLiked.value = !_isLiked.value
        
        viewModelScope.launch {
            try {
                val request = ToggleVideoLikeRequest.newBuilder()
                    .setVideoId(video.id)
                    .setIsLiked(_isLiked.value)
                    .build()
                
                val response = grpcClient.toggleVideoLike(request)
                if (!response.success) {
                    // Revert if failed
                    _isLiked.value = !_isLiked.value
                }
            } catch (e: Exception) {
                // Revert if failed
                _isLiked.value = !_isLiked.value
            }
        }
    }
    
    fun toggleFollow() {
        val video = _currentVideo.value ?: return
        
        _isFollowing.value = !_isFollowing.value
        
        viewModelScope.launch {
            try {
                val request = ToggleFollowRequest.newBuilder()
                    .setUserId(video.author.userId)
                    .setIsFollowing(_isFollowing.value)
                    .build()
                
                val response = grpcClient.toggleFollow(request)
                if (!response.success) {
                    // Revert if failed
                    _isFollowing.value = !_isFollowing.value
                }
            } catch (e: Exception) {
                // Revert if failed
                _isFollowing.value = !_isFollowing.value
                }
        }
    }
    
    fun showComments() {
        _showComments.value = true
    }
    
    fun hideComments() {
        _showComments.value = false
    }
    
    fun showShare() {
        _showShare.value = true
    }
    
    fun hideShare() {
        _showShare.value = false
    }
    
    fun refreshVideos() {
        currentPage = 0
        _videos.value = emptyList()
        _currentVideoIndex.value = 0
        loadVideos()
    }
    
    fun clearError() {
        _error.value = null
    }
    
    // MARK: - Private Methods
    private fun updateVideoState() {
        val video = _currentVideo.value ?: return
        
        // Update like and follow state based on current video
        _isLiked.value = video.isLiked
        _isFollowing.value = video.author.isFollowing
    }
}

// MARK: - Data Models
enum class VideoTab(
    val value: String,
    val displayName: String
) {
    FOR_YOU("for_you", "For You"),
    TRENDING("trending", "Trending");
    
    val grpcType: xyz.sonet.app.grpc.proto.VideoTab
        get() = when (this) {
            FOR_YOU -> xyz.sonet.app.grpc.proto.VideoTab.VIDEO_TAB_FOR_YOU
            TRENDING -> xyz.sonet.app.grpc.proto.VideoTab.VIDEO_TAB_TRENDING
        }
}

data class VideoItem(
    val id: String,
    val videoUrl: String,
    val thumbnailUrl: String,
    val caption: String,
    val author: UserProfile,
    val likeCount: Int,
    val commentCount: Int,
    val shareCount: Int,
    val viewCount: Int,
    val duration: Long,
    val isLiked: Boolean,
    val isFollowing: Boolean,
    val timestamp: Date,
    val hashtags: List<String>,
    val music: VideoMusic?
) {
    companion object {
        fun from(grpcVideo: xyz.sonet.app.grpc.proto.VideoItem): VideoItem {
            return VideoItem(
                id = grpcVideo.videoId,
                videoUrl = grpcVideo.videoUrl,
                thumbnailUrl = grpcVideo.thumbnailUrl,
                caption = grpcVideo.caption,
                author = UserProfile.from(grpcVideo.author),
                likeCount = grpcVideo.likeCount.toInt(),
                commentCount = grpcVideo.commentCount.toInt(),
                shareCount = grpcVideo.shareCount.toInt(),
                viewCount = grpcVideo.viewCount.toInt(),
                duration = grpcVideo.duration,
                isLiked = grpcVideo.isLiked,
                isFollowing = grpcVideo.author.isFollowing,
                timestamp = Date(grpcVideo.timestamp.seconds * 1000),
                hashtags = grpcVideo.hashtagsList,
                music = if (grpcVideo.hasMusic) VideoMusic.from(grpcVideo.music) else null
            )
        }
    }
}

data class VideoMusic(
    val id: String,
    val title: String,
    val artist: String,
    val audioUrl: String,
    val duration: Long
) {
    companion object {
        fun from(grpcMusic: xyz.sonet.app.grpc.proto.VideoMusic): VideoMusic {
            return VideoMusic(
                id = grpcMusic.musicId,
                title = grpcMusic.title,
                artist = grpcMusic.artist,
                audioUrl = grpcMusic.audioUrl,
                duration = grpcMusic.duration
            )
        }
    }
}

data class VideoComment(
    val id: String,
    val text: String,
    val author: UserProfile,
    val likeCount: Int,
    val isLiked: Boolean,
    val timestamp: Date,
    val replies: List<VideoComment>
) {
    companion object {
        fun from(grpcComment: xyz.sonet.app.grpc.proto.VideoComment): VideoComment {
            return VideoComment(
                id = grpcComment.commentId,
                text = grpcComment.text,
                author = UserProfile.from(grpcComment.author),
                likeCount = grpcComment.likeCount.toInt(),
                isLiked = grpcComment.isLiked,
                timestamp = Date(grpcComment.timestamp.seconds * 1000),
                replies = grpcComment.repliesList.map { from(it) }
            )
        }
    }
}

File: sonet-ios/Sonet/NativeApp/ViewModels/VideoFeedViewModel.swift
import Foundation
import Combine
import SwiftUI

@MainActor
class VideoFeedViewModel: ObservableObject {
    
    // MARK: - Published Properties
    @Published var videos: [VideoItem] = []
    @Published var currentVideoIndex = 0
    @Published var selectedTab: VideoTab = .forYou
    @Published var isLoading = false
    @Published var error: String?
    @Published var isLiked = false
    @Published var isFollowing = false
    @Published var showComments = false
    @Published var showShare = false
    @Published var currentVideo: VideoItem?
    
    // MARK: - Private Properties
    private let grpcClient: SonetGRPCClient
    private var cancellables = Set<AnyCancellable>()
    private var currentPage = 0
    private let pageSize = 10
    
    // MARK: - Initialization
    init(grpcClient: SonetGRPCClient) {
        self.grpcClient = grpcClient
        loadVideos()
    }
    
    // MARK: - Public Methods
    func loadVideos() {
        Task {
            isLoading = true
            error = nil
            
            do {
                let response = try await grpcClient.getVideos(
                    tab: selectedTab.grpcType,
                    page: currentPage,
                    pageSize: pageSize
                )
                
                let newVideos = response.videos.map { VideoItem(from: $0) }
                
                if currentPage == 0 {
                    videos = newVideos
                } else {
                    videos.append(contentsOf: newVideos)
                }
                
                if !videos.isEmpty {
                    currentVideo = videos[0]
                }
                
                currentPage += 1
            } catch {
                self.error = "Failed to load videos: \(error.localizedDescription)"
            }
            
            isLoading = false
        }
    }
    
    func switchTab(_ tab: VideoTab) {
        selectedTab = tab
        currentPage = 0
        videos.removeAll()
        currentVideoIndex = 0
        loadVideos()
    }
    
    func nextVideo() {
        guard currentVideoIndex < videos.count - 1 else {
            // Load more videos if needed
            loadVideos()
            return
        }
        
        currentVideoIndex += 1
        currentVideo = videos[currentVideoIndex]
        updateVideoState()
    }
    
    func previousVideo() {
        guard currentVideoIndex > 0 else { return }
        
        currentVideoIndex -= 1
        currentVideo = videos[currentVideoIndex]
        updateVideoState()
    }
    
    func selectVideo(at index: Int) {
        guard index >= 0 && index < videos.count else { return }
        
        currentVideoIndex = index
        currentVideo = videos[index]
        updateVideoState()
    }
    
    func toggleLike() {
        guard let video = currentVideo else { return }
        
        isLiked.toggle()
        
        Task {
            do {
                let request = ToggleVideoLikeRequest()
                request.videoId = video.id
                request.isLiked = isLiked
                
                let response = try await grpcClient.toggleVideoLike(request: request)
                if !response.success {
                    // Revert if failed
                    isLiked.toggle()
                }
            } catch {
                // Revert if failed
                isLiked.toggle()
            }
        }
    }
    
    func toggleFollow() {
        guard let video = currentVideo else { return }
        
        isFollowing.toggle()
        
        Task {
            do {
                let request = ToggleFollowRequest()
                request.userId = video.author.userId
                request.isFollowing = isFollowing
                
                let response = try await grpcClient.toggleFollow(request: request)
                if !response.success {
                    // Revert if failed
                    isFollowing.toggle()
                }
            } catch {
                // Revert if failed
                isFollowing.toggle()
            }
        }
    }
    
    func showComments() {
        showComments = true
    }
    
    func showShare() {
        showShare = true
    }
    
    func refreshVideos() {
        currentPage = 0
        videos.removeAll()
        currentVideoIndex = 0
        loadVideos()
    }
    
    // MARK: - Private Methods
    private func updateVideoState() {
        guard let video = currentVideo else { return }
        
        // Update like and follow state based on current video
        isLiked = video.isLiked
        isFollowing = video.author.isFollowing
    }
}

// MARK: - Data Models
enum VideoTab: String, CaseIterable {
    case forYou = "for_you"
    case trending = "trending"
    
    var displayName: String {
        switch self {
        case .forYou: return "For You"
        case .trending: return "Trending"
        }
    }
    
    var grpcType: xyz.sonet.app.grpc.proto.VideoTab {
        switch self {
        case .forYou: return .VIDEO_TAB_FOR_YOU
        case .trending: return .VIDEO_TAB_TRENDING
        }
    }
}

struct VideoItem: Identifiable {
    let id: String
    let videoUrl: String
    let thumbnailUrl: String
    let caption: String
    let author: UserProfile
    let likeCount: Int
    let commentCount: Int
    let shareCount: Int
    let viewCount: Int
    let duration: TimeInterval
    let isLiked: Bool
    let isFollowing: Bool
    let timestamp: Date
    let hashtags: [String]
    let music: VideoMusic?
    
    init(from grpcVideo: xyz.sonet.app.grpc.proto.VideoItem) {
        self.id = grpcVideo.videoId
        self.videoUrl = grpcVideo.videoUrl
        self.thumbnailUrl = grpcVideo.thumbnailUrl
        self.caption = grpcVideo.caption
        self.author = UserProfile(from: grpcVideo.author)
        self.likeCount = Int(grpcVideo.likeCount)
        self.commentCount = Int(grpcVideo.commentCount)
        self.shareCount = Int(grpcVideo.shareCount)
        self.viewCount = Int(grpcVideo.viewCount)
        self.duration = TimeInterval(grpcVideo.duration)
        self.isLiked = grpcVideo.isLiked
        self.isFollowing = grpcVideo.author.isFollowing
        self.timestamp = Date(timeIntervalSince1970: TimeInterval(grpcVideo.timestamp.seconds))
        self.hashtags = grpcVideo.hashtagsList
        self.music = grpcVideo.hasMusic ? VideoMusic(from: grpcVideo.music) : nil
    }
}

struct VideoMusic {
    let id: String
    let title: String
    let artist: String
    let audioUrl: String
    let duration: TimeInterval
    
    init(from grpcMusic: xyz.sonet.app.grpc.proto.VideoMusic) {
        self.id = grpcMusic.musicId
        self.title = grpcMusic.title
        self.artist = grpcMusic.artist
        self.audioUrl = grpcMusic.audioUrl
        self.duration = TimeInterval(grpcMusic.duration)
    }
}

struct VideoComment: Identifiable {
    let id: String
    let text: String
    let author: UserProfile
    let likeCount: Int
    let isLiked: Bool
    let timestamp: Date
    let replies: [VideoComment]
    
    init(from grpcComment: xyz.sonet.app.grpc.proto.VideoComment) {
        self.id = grpcComment.commentId
        self.text = grpcComment.text
        self.author = UserProfile(from: grpcComment.author)
        self.likeCount = Int(grpcComment.likeCount)
        self.isLiked = grpcComment.isLiked
        self.timestamp = Date(timeIntervalSince1970: TimeInterval(grpcComment.timestamp.seconds))
        self.replies = grpcComment.repliesList.map { VideoComment(from: $0) }
    }
}